Add ctx to `beforeSSR` and return 404 error for 404 pages

michalczaplinski opened the pull request:
#### Description of what you did:
Add `ctx` to `beforeSSR` and return correct HTTP  status code in frontity for corresponding errors from a server.

- [x] Add `ctx` to `beforeSSR`.
- [x] Update the docs (we can add it here https://docs.frontity.org/learning-frontity/actions#actions-triggered-by-frontity)

fixes: #306 
fixes: #264 

[Documentation](https://app.gitbook.com/@frontity-tech/s/frontity/~/diff/drafts/-M00Ls29V3QLo_LSrcue/@drafts)

#### My PR is a:
- 🔝 Improvement

#### Is the PR ready to be merged?
- ✅ Done

michalczaplinski committed:
tiny-router, core: add ctx to beforeSSR

michalczaplinski committed:
tiny-router: update test with updated types

michalczaplinski committed:
tiny-router: Add correct source param to state

michalczaplinski committed:
tiny-router: remove package-lock

luisherranz commented on packages/core/src/server/index.tsx:
```diff
@@ -79,7 +79,7 @@ export default ({ packages }): ReturnType<Koa["callback"]> => {
     await Promise.all(
       Object.values(store.actions).map(({ beforeSSR }) => {
-        if (beforeSSR) return beforeSSR();
+        if (beforeSSR) return beforeSSR(ctx);
```
I would make this an object, just in case we need to add more stuff in the future:
```js
beforeSSR({ ctx });
```

luisherranz commented on packages/tiny-router/src/actions.ts:
```diff
@@ -43,16 +44,24 @@ export const init: TinyRouter["actions"]["router"]["init"] = ({
   actions
-}) => {
+}) => async (
+  ctx: Koa.ParameterizedContext<Koa.DefaultState, Koa.DefaultContext>
```
For TypeScript I think I would expose a `ServerAction` in `"@frontity/types"` with our own `Context` type because it will include Koa's context but also more things in the future (like `ctx.settings`, `ctx.template` and so on...).

```ts
import { ServerAction } from "frontity/types";

actions: {
  myPackage: {
    beforeSSR: ServerAction<MyPackage>;
  ...
}
```

That way, people can launch other server actions from the `beforeSSR` one and organize their code better:

```ts
actions: {
  myPackage: {
    beforeSSR: ServerAction<MyPackage>;
    doSomethingWithContext: ServerAction<MyPackage>;
    doSomethingElseWithContext: ServerAction<MyPackage>;
  ...
}
```
```js
actions: {
  myPackage: {
    beforeSSR: ({ actions }) => async ({ ctx }) => {
      await Promise.all([
        actions.myPackage.doSomethingWithContext({ ctx }),
        actions.myPackage.doSomethingElseWithContext({ ctx })
      ]);
    }
  }
}
```

luisherranz commented on packages/tiny-router/src/index.ts:
```diff
@@ -1,13 +1,15 @@
       autoFetch: true
-    }
+    },
+    source: WPSource().state.source // TODO: notr sure if that's correct
```
It's not correct. The state from other packages is available on runtime.

michalczaplinski replied:
gotcha 👍

luisherranz reviewed (commented):
I know this is not ready for review yet, but I wanted to give you my impressions :)

luisherranz commented on packages/tiny-router/src/actions.ts:
```diff
@@ -43,16 +44,24 @@ export const init: TinyRouter["actions"]["router"]["init"] = ({
+
+    const data = state.source.get(state.router.link);
+    if (data.isError) {
+      ctx.status = data.errorStatus;
```
Hey @michalczaplinski, I guess you're solving this issue in this PR as well, right? https://github.com/frontity/frontity/issues/264

If so, could you please link the issue, assign it to you and move it to in progress? Thanks!

michalczaplinski replied:
yup, done.

michalczaplinski committed:
wip: commiting for luis to take a look

luisherranz committed:
Fix wp-source errors e2e tests

They were failing because cy.request and cy.visit throw when the error
codes are different than 2xx or 3xx.

luisherranz commented:
It looks like the problem in the e2e test was that Cypress throws for `cy.request` and `cy.visit` unless you add a `failOnStatusCode: false` option. Check:
- https://docs.cypress.io/api/commands/visit.html#Arguments
- https://docs.cypress.io/api/commands/request.html#Arguments

----

Apart from that, I was not able to reproduce the problem you describe in [this video](https://www.loom.com/share/6eb65393bd6d43259c365bab2ff2489c). For me, `await actions.source.fetch(...)` works fine. Take a look at the e2e test: https://github.com/frontity/frontity/blob/917d34bdafc3bca9a969ca4165e70205de42ef98/e2e/packages/wp-source-errors/src/index.tsx#L32-L34

luisherranz committed:
Fix tiny-router types

Because state.source.get was missing.

luisherranz committed:
Merge branch 'dev' into koa-context

luisherranz commented:
I have fixed the types of `tiny-router` which were missing `state.source.get` and I have merged `dev` because this branch was out of sync.

michalczaplinski commented on e2e/project/frontity.settings.ts:
```diff
@@ -35,11 +35,7 @@ const settings: Settings = [
-      "@frontity/wp-source",
-      "@frontity/tiny-router"
-    ]
+    packages: ["e2e-wp-source-errors", "@frontity/wp-source"]
```
@luisherranz I'm a bit confused. Why did you remove the `tiny-router`? 

I've added it because I needed to test that the `beforeSSR` *in the* tiny-router package adds the correct error message.

luisherranz replied:
I thought that it was not needed to test that we can change the `ctx.status` from a `beforeSSR` method... but it's fine. Add it back if you want.

michalczaplinski commented on e2e/packages/wp-source-errors/src/index.tsx:
```diff
@@ -28,8 +28,10 @@ const wpSourceErrors: WpSourceErrors = {
         }
       },
-      beforeSSR: async ({ state, actions }) => {
+      beforeSSR: ({ state, actions }) => async ctx => {
```
@luisherranz This was not relevant for my tests, because I needed to rely on the `beforeSSR` of the `tiny-router`. 

The strange thing is that with your changes, I can comment out the content of the `beforeSSR` in this file and the tests pass **and** I get the correct errors when I navigate to the page manually with my browser.

luisherranz replied:
Oh, do you mean that the tests pass without any code? Maybe the tests are wrong... I don't know. Could you please check them out?

michalczaplinski replied:
I meant that they pass because of this: https://github.com/frontity/frontity/pull/309/files#diff-d8058b9e7fa71c069f1e5c2f6a255444L50-R65

We set the correct status code in the `beforeSSR` action of the `tiny-router`.

michalczaplinski reviewed (commented):
@luisherranz I don't fully understand why this is working now and was not working before. 

What I did on top of your changes: 
1. I re-added the `tiny-router` to the `wp-source-errors` [here](https://github.com/frontity/frontity/pull/309/commits/917d34bdafc3bca9a969ca4165e70205de42ef98#diff-d35f28fb9a23785741a203fa518e1450R38)
2. I commented out the `beforeSSR` method that you added [here](https://github.com/frontity/frontity/pull/309/commits/917d34bdafc3bca9a969ca4165e70205de42ef98#diff-14adba8e1fc2b93ce0a92fb4a9625e9eR31)

The test pass now (I know that I was missing the `failOnStatusCode: false` as you pointed out, but regardless of that koa was returning 200 when it should have returned an error)

michalczaplinski committed:
tiny-router: fix e2e tests

michalczaplinski committed:
tiny-router: Remove not needed wp-source from package.json

luisherranz commented:
I am a bit confused about your messages... Is this working? Are the tests correct? Do you need more help?

michalczaplinski commented:
Nah, I'm fine now. I still don't know why I was having the issue that I described, but since I cant reproduce it now, I'm just gonna move on.

luisherranz commented:
Great! 👍

michalczaplinski committed:
tiny-router: Add the correct types

michalczaplinski committed:
types: fix the ServerAction type

michalczaplinski commented on packages/types/src/action.ts:
```diff
@@ -42,3 +74,11 @@ export type AsyncAction<Pkg extends Package, Input = null> = [Input] extends [
     }) => (args: Input) => Promise<void>;
+
+type Context = Koa.ParameterizedContext<Koa.DefaultState, Koa.DefaultContext>;
+
```
@luisherranz I think this is correct? We have to support every use case like the following, right?

```js
// SYNC
// 1. sync, no params
action: ({state, ...rest }) => {}

// 2. sync, with params
action: ({state, ...rest }) => params =>  {}

// ASYNC
// 3. async, no params
action: async ({state, ...rest }) =>  {}

// 4. async, with params
action: ({state, ...rest }) => async (params) => {}
```

michalczaplinski replied:
oh, I realized that `beforeSSR` is supposed to always be async (at least according to the docs)

michalczaplinski replied:
nevermind, that's not true. It can be both.

michalczaplinski commented on packages/tiny-router/types.ts:
```diff
@@ -8,6 +8,9 @@ interface TinyRouter extends Router {
     };
+    source?: {
+      get?: Source["state"]["source"]["get"];
+    };
```
Added cause we use `state.source.get()` inside the `beforeSSR()` in tiny-router

michalczaplinski committed:
Add changeset

michalczaplinski committed:
tiny-router: fix TS & use magic of type inference

michalczaplinski committed:
Add changest for Koa's context

luisherranz committed:
Remove internal packages changeset

We don't need to use changeset for internal packages, only for
external ones.

luisherranz committed:
Remove unnecessary eslint disable

luisherranz committed:
Remove unnecessary devDependency

luisherranz committed:
Remove unnecessary TypeScript declaration

luisherranz committed:
Remove unnecessary optional type

luisherranz committed:
Remove unnecessay import

luisherranz committed:
Move autoFetch warning to correct place

The else was affecting to the data.isError if, instead of to the
actions.source check

luisherranz committed:
Switch to frontity warn instead of console.warn

luisherranz committed:
Refactor test file

Move each action to its own describe and add mocks for source state and
actions

luisherranz committed:
Export context

I think that we can leave ServerAction as both async and requiring a
Context for the moment.

luisherranz committed:
Add unit tests for beforeSSR

luisherranz committed:
Merge branch 'dev' into koa-context

luisherranz reviewed (approved):
This is great. Awesome work @michalczaplinski!

By the way, I've added some missing unit tests and removed some things that were not needed. Check my commits and their messages for more information :)

luisherranz committed:
Only change context status if a source pkg exists

luisherranz commented:
Docs are great too so I'm going to merge this :)

michalczaplinski commented on packages/types/src/action.ts:
```diff
@@ -75,12 +77,9 @@ export type AsyncAction<Pkg extends Package, Input = null> = [Input] extends [
-  | Action<Pkg, Context>
-  | AsyncAction<Pkg>
-  | AsyncAction<Pkg, Context>;
+export type ServerAction<Pkg> = AsyncAction<Pkg, { ctx: Context }>;
```
@luisherranz You wrote in the commit message: 
> I think that we can leave ServerAction as both async and requiring a Context for the moment.

I thought that it all 4 should be possible:
- sync, no `ctx`
- sync, with `ctx`
- async, no `ctx`
- async, with `ctx`
But with the type that you added, TS will only allow "async with ctx". Is there something I'm missing?

See the comment above for more details.

luisherranz replied:
Yes, let's leave it like that for the moment. If people want to use other combinations in their packages they can use `Action` and `AsyncAction` when defining the types of their `beforeSSR` action.
